doc/TAM/SAI-Proposal-TAM-stream-telemetry.md

```diff
@@ -0,0 +1,906 @@
+        SAI_TAM_TELEMETRY_ATTR_REPORTING_TYPE SAI_TAM_REPORTING_TYPE_COUNT_BASED
+        SAI_TAM_TELEMETRY_ATTR_REPORTING_BULK_SIZE bulk_count "STREAM_TELEMETRY_PROFILE:profile_name[bulk_count] on Config DB"
+    }
+    counter_subscription[TAM_counter_subscription] {
+        SAI_ID SAI_VALUE "Comments"
+        SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_TEL_TYPE sai_tam_tel_type_obj
+        SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_OBJECT_ID port_obj
+        SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STAT_ID SAI_PORT_STAT_IF_IN_OCTETS "A stats in sai_port_stat_t"
+        SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_LABEL index "Element ID of the object in the IPFIX template"
+        SAI_TAM_COUNTER_SUBSCRIPTION_ATTR_STATS_MODE SAI_STATS_MODE_READ "Set read counter mode"
+    }
+    TAM[TAM] {
+        SAI_ID SAI_VALUE "Comments"
+        SAI_TAM_ATTR_TELEMETRY_OBJECTS_LIST sai_tam_telemetry_obj
+        SAI_TAM_ATTR_TAM_BIND_POINT_TYPE_LIST SAI_TAM_BIND_POINT_TYPE_PORT
+    }
+    switch[Switch] {
+        SAI_ID SAI_VALUE "Comments"
+        SAI_SWITCH_ATTR_TAM_OBJECT_ID sai_tam_obj
+        SAI_SWITCH_ATTR_TAM_CACHE_COUNT cache_count
+    }
+
+    host_table_entry |o--|| hostif: binds
+    host_table_entry |o--|| hostif_trap: binds
+    hostif_trap |o--|| hostif_trap_group: binds
+    collector |o--|| hostif_trap: binds
+    collector |o--|| transport: binds
+    telemetry_type |o--|| report: binds
+    telemetry |o--o| telemetry_type: binds
+    telemetry }o--o{ collector: binds
+    counter_subscription }o--|| telemetry_type: binds
+    TAM |o--o| telemetry: binds
+    switch |o..o{ TAM: binds
+```
+
+| Object Type              | Scope                        |
+| ------------------------ | ---------------------------- |
+| HOSTIF                   | Global                       |
+| HOSTIF_trap_group        | Global                       |
+| HostIF_user_defined_trap | Global                       |
+| HOSTIF_table_entry       | Global                       |
+| TAM_transport            | Global                       |
+| TAM_collector            | Global                       |
+| TAM                      | per STREAM_TELEMETRY profile |
+| TAM_telemetry            | per STREAM_TELEMETRY profile |
+| TAM_telemetry_type       | per STREAM_TELEMETRY profile |
+| TAM_report               | per STREAM_TELEMETRY profile |
+| TAM_counter_subscription | per stats of object          |
+
+#### Initialize TAM cache count for Switch
+
+``` c++
+/**
+ * @brief Attribute Id in sai_set_switch_attribute() and
+ * sai_get_switch_attribute() calls.
+ */
+typedef enum _sai_switch_attr_t
+{
+    // ...
+
+    /**
+     * @brief Tam telemetry cache count
+     *
+     * If the collector isn't ready to receive the report, this value indicates how many
+     * reports that can be cached. 0 means no cache which is the default behavior.
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_ONLY
+     * @default 0
+     */
+    SAI_SWITCH_ATTR_TAM_CACHE_COUNT,

```

Comment:
What is the caching semantics? Say the collector is not ready, and more reports are generated than the cache capacity, are the first 'n' or the last 'n' cached?

Please also explain why this is required?

Reply:
Thanks for your comments. Update the comments.

Reply:
We hope it's possible to collect the counters during the collector restarts after the collector crash.